File: Week_03/G20200343030632/Leetcode_50_632.java
class Solution {
    public double myPow(double x, int n) {
        if (n < 0) {
            x = 1 / x;
            n = -n;
        }
        return helper(x, n);
    }

    private double helper(double x, int n) {
        if (n == 0) {
            return 1.0;
        }
        double res = helper(x, n / 2);
        if (n % 2 == 0) {
            return res * res;
        } else {
            return res * res * x;
        }
    }
}

File: Week_03/G20200343030632/Leetcode_50_632.py
class Solution:
    def myPow(self, x: float, n: int) -> float:
        if n < 0:
            x, n = 1 / x, -1 * n
        res = self.helper(x, n)            
        return res
    
    def helper(self, x: float, n: int) -> float:
        if n == 0:
            return 1.0
        res = self.helper(x, n // 2)
        if n % 2 == 0:
            return res * res
        else:
            return res * res * x